KEP-4330: extend version information with more detailed version fields

yongruilin opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature

#### What this PR does / why we need it:
- Add new version fields to version.Info struct:
  * BinaryMajor and BinaryMinor to track binary version
  * MinCompatibilityMajor and MinCompatibilityMinor for compatibility tracking
- Update related code to populate and use these new fields
- Improve version information documentation and OpenAPI generation
- Modify version routes and documentation to reflect new version information structure

Part of [KEP-4330](https://github.com/kubernetes/enhancements/issues/4330) to extend the [/version endpoint](https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/4330-compatibility-versions/README.md#version-introspection)

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #https://github.com/kubernetes/kubernetes/issues/129969

#### Special notes for your reviewer:
This would also require some client-side change to separate to consume binary version and compatibility version.
#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Update /version response to report binary version information separate from compatibility version
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/4330-compatibility-versions/README.md
```

yongruilin committed:
feat: extend version information with more detailed version fields

- Add new version fields to version.Info struct:
  * EmulationMajor and EmulationMinor to track emulated version
  * MinCompatibilityMajor and MinCompatibilityMinor for compatibility tracking
- Update related code to populate and use these new fields
- Improve version information documentation and OpenAPI generation
- Modify version routes and documentation to reflect new version information structure

yongruilin commented:
/cc @siyuanfoundation @jpbetz 
This may require a follow up change on the client side: kubectl, client-go, kubeadm

neolit123 commented on staging/src/k8s.io/apimachinery/pkg/version/types.go:
```diff
@@ -20,15 +20,25 @@ package version
-	GoVersion    string `json:"goVersion"`
-	Compiler     string `json:"compiler"`
-	Platform     string `json:"platform"`
+	// Major is the major version of the effective version
+	Major string `json:"major"`
+	// Minor is the minor version of the effective version
+	Minor string `json:"minor"`
```
trying to understand the implications here.
so if one passes --version and  the new compatibility-version flags on the cli they will get varying output?
if so, this pr touches not only the /version endpoints.

neolit123 commented on staging/src/k8s.io/apimachinery/pkg/version/types.go:
```diff
@@ -20,15 +20,25 @@ package version
+	Major string `json:"major"`
+	// Minor is the minor version of the effective version
+	Minor string `json:"minor"`
+	// BinaryMajor is the major version of the binary
+	BinaryMajor string `json:"binaryMajor"`
+	// BinaryMinor is the minor version of the binary
+	BinaryMinor string `json:"binaryMinor"`
```
since this is technically an API that is in active use currently. i think this should be done the other way around.
Major and Minor remain the original binary versions.
CompatibilityMajor and CompatibilityMinor are the new introduced fields.

siyuanfoundation replied:
I agree. It is better to use `CompatibilityMajor` and `CompatibilityMinor` instead of changing the meaning of existing api.

yongruilin replied:
I thought the intention from the KEP was to always report the effective version and separate the binary version. https://github.com/kubernetes/enhancements/pull/4395#discussion_r1482085465

But I agree it is a breaking change as this PR change the meaning of current fields being widely used. We can introduce the new fields to represent the emulated versions and min compatibility versions, but meanwhile rephrase the KEP. wdyt?

siyuanfoundation replied:
I agree. We can make an update to the KEP. 
@jpbetz wdyt?

yongruilin replied:
To make a more compatible change, I've updated the PR to add `EmulationMajor/Minor` instead of changing the current meaning of "Major/Minor".

BenTheElder replied:
We also have `GitVersion` exposed in the API here, I know we've discussed this area before but I think it will be confusing to have two concepts of the version of the binary on top of the emulated version + emulation bounds.

How do expect clients to use these? You could argue for compat reasons it's better to lie about the version to be the emulated version if we think that will change the behavior?

Also worth noting: there is one additional flag already to override the reported version which controls some or all of these (`--version` for api-server only IIRC).

It doesn't look like https://github.com/kubernetes/enhancements/pull/4395#discussion_r1482085465 got a lot of discussion about which way to report this, cc @jpbetz @liggitt @aojea @jeremyrickard

aojea replied:
IIUIC this changes moves from
```json
{
  "major": "1",
  "minor": "30",
  "gitVersion": "v1.30.0",
  "gitCommit": "<something>",
  "gitTreeState": "clean",
  "buildDate": "2024-03-30T06:36:32Z",
  "goVersion": "go1.21.something",
  "compiler": "gc",
  "platform": "linux/arm64"
}
```

to

```json
{
  "major": "1",
  "minor": "30",
  "emulationMajor": "1",
  "emulationMinor": "29",
  "minCompatibilityMajor": "1",
  "minCompatibilityMinor": "29",
  "gitVersion": "v1.30.0",
  "gitCommit": "<something>",
  "gitTreeState": "clean",
  "buildDate": "2024-03-30T06:36:32Z",
  "goVersion": "go1.21.something",
  "compiler": "gc",
  "platform": "linux/arm64"
}
```

since major and minor retain the same meaning and the changes are additive fields 
```
  "emulationMajor": "1",
  "emulationMinor": "29",
  "minCompatibilityMajor": "1",
  "minCompatibilityMinor": "29",
```

this looks a compatible API change.

>  You could argue for compat reasons it's better to lie about the version to be the emulated version if we think that will change the behavior?

I don't think that emulated version means as "exactly that version", IMHO is more as in "using the same feature gated code as in that version", and having fields that represent exactly that seems better for the integrations that use that feature

jpbetz replied:
> I agree. We can make an update to the KEP. @jpbetz wdyt?

Yes, I just looked at the KEP, and we don't want to be changing existing fields, so feel free to update it to match what we decide here.

Having new "emulation" version fields (and compatibility versions fields if/when we implement that aspect of the KEP) seems right.  

Would the "emulation" version fields be set only when the `--emulation-version` flag is set?

siyuanfoundation replied:
It is probably better to have "emulation" version fields set even when the --emulation-version flag is not set, so that users of the emulation version do not need to do the conversion.

yongruilin replied:
Created https://github.com/kubernetes/enhancements/pull/5180 for updating the KEP.

BenTheElder replied:
> I don't think that emulated version means as "exactly that version", IMHO is more as in "using the same feature gated code as in that version", and having fields that represent exactly that seems better for the integrations that use that feature instead of changing the meaning, that will impact all existing consumers of that feature.

Yeah, that makes sense. And even without emulation-version, you can have a component setting feature gates to non-default values anyhow. The binary version is still the primary version.

> I agree with adding new fields for the emulation and compatibility versions.

+1

> It is probably better to have "emulation" version fields set even when the --emulation-version flag is not set, so that users of the emulation version do not need to do the conversion.

+1, the logic on defaulting should be server-side

neolit123 commented on staging/src/k8s.io/apimachinery/pkg/version/types.go:
```diff
@@ -20,15 +20,25 @@ package version
+	// BinaryMinor is the minor version of the binary
+	BinaryMinor string `json:"binaryMinor"`
+	// MinCompatibilityMajor is the major version of the minimum compatibility version
+	MinCompatibilityMajor string `json:"minCompatibilityMajor"`
```
beter to have the compat major. compat major would only be needed if k8s eventually moves to 2.x which mean the last 1.x can be a compat major. having said that, a 2.x k8s would break a lot of things across the board.

siyuanfoundation replied:
I don't think we will support emulation across major releases...
Having said that, it is pretty trivial to add the compat major. So I am ok with both.

yongruilin commented:
/retest

siyuanfoundation commented:
/lgtm

/triage accepted

yongruilin commented:
/assign @deads2k

deads2k commented on staging/src/k8s.io/component-base/compatibility/version.go:
```diff
@@ -173,6 +177,29 @@ func (m *effectiveVersion) Validate() []error {
 
+// Info returns the version information of a component.
+// If the binary version is nil, it returns nil.
+func (m *effectiveVersion) Info() *apimachineryversion.Info {
```
Given that `apimachineryversion.Info`, `BinaryVersion`, `EmulationVersion`, `MinCompatibilityVersion`, `binaryVersion`, `emulationVersion`, `minCompatibilityVersion` are all footguns that can be conflicting because `version.Version` now includes all that information via its `.info`.  The interfaces need to be cleaned up here as part of the switch or perhaps `pkg/util/version/version.go` should be focused on a single version.  I'm flexible for a solution, but the potential for conflict needs to be eliminated.

siyuanfoundation replied:
good point. @yongruilin can we remove `apimachineryversion.Info` from
```
type Version struct {
	components    []uint
	semver        bool
	preRelease    string
	buildMetadata string
	info          apimachineryversion.Info
}
```
so that pkg/util/version/version.go is focused on a single version?

yongruilin replied:
Detached the info field from util/version. PTAL

yongruilin committed:
refactor: detach Info from apimachinery util version

- Remove `info` field from `Version` struct
- Modify `WithInfo` and `Info` methods to be deprecated
- Update version information retrieval to use base version info
- Simplify version information generation in compatibility tests
- Remove unnecessary version info passing in build and test scenarios

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/routes/version.go:
```diff
@@ -39,11 +39,11 @@ func (v Version) Install(c *restful.Container) {
 	versionWS.Route(
 		versionWS.GET("/").To(v.handleVersion).
-			Doc("get the code version").
-			Operation("getCodeVersion").
```
Why is this change needed?  I don't know enough about openapi offhand to know if this is semantically important, but it smells semantically important.  I wouldn't expect it to change unless we must change it.

deads2k commented:
Just the comment about the operation and it lgtm.

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/util/version/version.go:
```diff
@@ -456,27 +455,26 @@ func (v *Version) Compare(other string) (int, error) {
 }
 
-// WithInfo returns copy of the version object with requested info
+// Deprecated: WithInfo returns copy of the version object with requested info
```
```suggestion
// WithInfo returns copy of the version object with requested info
Deprecated: Use WithInfo() from effective version instead.
```

?

yongruilin replied:
I think `WithInfo(...)` should be deleted as it doesn't take any effect. It shouldn't be a breaking change?

yongruilin replied:
Ok, I've updated the comment.

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/util/version/version.go:
```diff
@@ -456,27 +455,26 @@ func (v *Version) Compare(other string) (int, error) {
 	return &result
 }
 
+// Deprecated: Use Info() from effective version instead.
```
```suggestion
Info returns the version information of a component.
// Deprecated: Use Info() from effective version instead.
```

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/version/types.go:
```diff
@@ -20,15 +20,25 @@ package version
+	// Minor is the minor version of the binary version
+	Minor string `json:"minor"`
+	// EmulationMajor is the major version of the emulation version
+	EmulationMajor string `json:"emulationMajor"`
+	// EmulationMinor is the minor version of the emulation version
+	EmulationMinor string `json:"emulationMinor"`
+	// MinCompatibilityMajor is the major version of the minimum compatibility version
+	MinCompatibilityMajor string `json:"minCompatibilityMajor"`
+	// MinCompatibilityMinor is the minor version of the minimum compatibility version
+	MinCompatibilityMinor string `json:"minCompatibilityMinor"`
```
Are these fields always set, or should we add "omitempty" ?

jpbetz commented on staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go:
```diff
@@ -278,7 +279,7 @@ func TestServerRunWithSNI(t *testing.T) {
 			config := setUp(t)
 			v := fakeVersion()
-			config.EffectiveVersion = basecompatibility.NewEffectiveVersionFromString(v.String(), "", "")
+			config.EffectiveVersion = basecompatibility.NewEffectiveVersionFromString(fmt.Sprintf("%s.%s", v.Major, v.Minor), "", "")
```
nit: It seems a bit silly to serialize to a string and then parse back.  Is there a reason we can't either pass the version or pass the major, minor parts?

yongruilin replied:
The "v" here is `info` object instead of a "version.Version"(I've updated the naming).
The fields are all just strings.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/server/options/serving_test.go:
```diff
@@ -372,7 +373,7 @@ func TestServerRunWithSNI(t *testing.T) {
 			}
 			if expected := &v; !reflect.DeepEqual(got, expected) {
-				t.Errorf("loopback client didn't get correct version info: expected=%v got=%v", expected, got)
+				t.Errorf("loopback client didn't get correct version info: expected=%#v got=%#v", *expected, *got)
```
We're _sure_ that `got` can never be nil here?

yongruilin replied:
Ah.. missed to change it back. Updated.

jpbetz reviewed (commented):
Minor cleanup suggestions.

jpbetz commented:
/approve
If anyone else wants to hop on this one, please do so soon, or throw on a hold

Leaving LGTM with @siyuanfoundation

jpbetz commented:
A few files here look like they need approval from @kubernetes/api-reviewers

yongruilin commented:
> A few files here look like they need approval from @kubernetes/api-reviewers

Looks like it needs approval from @kubernetes/api-approvers

thockin commented on staging/src/k8s.io/apimachinery/pkg/version/types.go:
```diff
@@ -20,15 +20,25 @@ package version
+	// Minor is the minor version of the binary version
+	Minor string `json:"minor"`
+	// EmulationMajor is the major version of the emulation version
+	EmulationMajor string `json:"emulationMajor,omitempty"`
```
Did we consider something like:

```
type Info struct {
    MajorMinor
    Emulation MajorMinor
    MinCompat MajorMinor
}

type MajorMinor struct {
    Major        string `json:"major"`
	Minor        string `json:"minor"`
}
```

?  It's sort of annoying to have another type, but I feel obligated to ask

yongruilin replied:
I did. It does look cleaner in this type, but requires a much larger scope of refactoring, including update the client side. We sure can follow up to do so but I don't think it worth the effort.

thockin reviewed (commented):
I don't know enough about this endpoint to challenge the design - the parts that need my approval seem consistent with the rest and I trust the oversight of that so:

/approve

yongruilin commented:
/test pull-kubernetes-integration
